Mathlib/Algebra/Order/Ring/Lemmas.lean

```diff
@@ -9,110 +9,150 @@ import Mathlib.Algebra.GroupWithZero.Defs
-    expressing that multiplication by positive elements on the left is strictly reverse monotone;
-* reverse reverse monotone right
-* * `ContravariantClass α>0 α (fun x y ↦ y * x) (≤)`, abbreviated `MulPosMonoRev α`,
-    expressing that multiplication by positive elements on the right is reverse monotone;
-* * `ContravariantClass α>0 α (fun x y ↦ y * x) (<)`, abbreviated `MulPosReflectLT α`,
-    expressing that multiplication by positive elements on the right is strictly reverse monotone.
+# Monotonicity of multiplication by positive elements
+
+This file defines typeclasses to reason about monotonicity of the operations
+* `b ↦ a * b`, "left multiplication"
+* `a ↦ a * b`, "right multiplication"
+
+We use eight typeclasses to encode the various properties we care about for those two operations.
+These typeclasses are meant to be mostly internal to this file, to set up each lemma in the
+appropriate generality.
+
+Less granular typeclasses like `OrderedAddCommMonoid`, `LinearOrderedField`` should be enough for
+most purposes, and the system is set up so that they imply the correct granular typeclasses here.
+If those are enough for you, you may stop reading here! Else, beware that what follows is a bit
+technical.
+
+## Definitions
+
+In all that follows, `α` is an orders which has a `0` and a multiplication. Note however that we do
+not use lawfulness of this action in most of the file. Hence `*` should be considered here as a
+mostly arbitrary function `α → α → α`.
+
+We use the following four typeclasses to reason about left multiplication (`b ↦ a * b`):
+* `PosMulMono`: If `a ≥ 0`, then `b₁ ≤ b₂ → a * b₁ ≤ a * b₂`.
+* `PosMulStrictMono`: If `a > 0`, then `b₁ < b₂ → a * b₁ < a * b₂`.
+* `PosMulReflectLT`: If `a ≥ 0`, then `a * b₁ < a * b₂ → b₁ < b₂`.
+* `PosMulMonoRev`: If `a > 0`, then `a * b₁ ≤ a * b₂ → b₁ ≤ b₂`.
+
+We use the following four typeclasses to reason about right multiplication (`a ↦ a * b`):
+* `MulPosMono`: If `b ≥ 0`, then `a₁ ≤ a₂ → a₁ * b ≤ a₂ * b`.
+* `MulPosStrictMono`: If `b > 0`, then `a₁ < a₂ → a₁ * b < a₂ * b`.
+* `MulPosReflectLT`: If `b ≥ 0`, then `a₁ * b < a₂ * b → a₁ < a₂`.
+* `MulPosMonoRev`: If `b > 0`, then `a₁ * b ≤ a₂ * b → a₁ ≤ a₂`.
+
+As `α` gets more and more structure, those typeclasses end up being equivalent. The commonly used
+implications are:
+* `PosMulStrictMono → PosMulMono`, `MulPosStrictMono → MulPosMono`,
+  `PosMulMonoRev → PosMulReflectLT`, `MulPosMonoRev → MulPosReflectLT` when `α` is a partial order.
+* `PosMulStrictMono → PosMulMonoRev`, `MulPosStrictMono → MulPosMonoRev` when `α` is a linear order.
+* `PosMulMono → MulPosMono`, `PosMulStrictMono → MulPosStrictMono`, `PosMulMonoRev → MulPosMonoRev`,
+  `PosMulReflectLT → MulPosReflectLT` when the multiplication of `α` is commutative.
+* `PosMulMono → MulPosMono`, `PosMulStrictMono → MulPosStrictMono` when `α` is an ordered ring.
+* `PosMulStrictMono → PosMulReflectLT`, `PosMulMono → PosMulMonoRev` when `α` is an ordered
+  semifield.
```

Comment:
It would be cool if we could display this as:
```mermaid

graph BT
  PosMulStrictMono --> PosMulMono
  MulPosStrictMono --> MulPosMono
  PosMulMonoRev --> PosMulReflectLT
  MulPosMonoRev -->|α is a partial order| MulPosReflectLT
  PosMulStrictMono --> PosMulMonoRev
  MulPosStrictMono -->|α is a linear order| MulPosMonoRev
  PosMulMono --> MulPosMono
  PosMulStrictMono --> MulPosStrictMono
  PosMulMonoRev --> MulPosMonoRev
  PosMulReflectLT -->|the multiplication of α is commutative| MulPosReflectLT
  PosMulMono --> MulPosMono
  PosMulStrictMono -->|α is an ordered ring| MulPosStrictMono
  PosMulStrictMono --> PosMulReflectLT
  PosMulMono -->|α is an ordered semifield| PosMulMonoRev
```

Reply:
```suggestion
* when `α` is a partial order: `PosMulStrictMono → PosMulMono`, `MulPosStrictMono → MulPosMono`,
  `PosMulMonoRev → PosMulReflectLT`, `MulPosMonoRev → MulPosReflectLT`.
* `PosMulStrictMono → PosMulMonoRev`, `MulPosStrictMono → MulPosMonoRev` when `α` is a linear order.
* `PosMulMono → MulPosMono`, `PosMulStrictMono → MulPosStrictMono`, `PosMulMonoRev → MulPosMonoRev`,
  `PosMulReflectLT → MulPosReflectLT` when the multiplication of `α` is commutative.
* `PosMulMono → MulPosMono`, `PosMulStrictMono → MulPosStrictMono` when `α` is an ordered ring.
* `PosMulStrictMono → PosMulReflectLT`, `PosMulMono → PosMulMonoRev` when `α` is an ordered
  semifield.
```
etc: I read this last clause as modifying only the last list item originally

Reply:
Be the module doc you want to see!